File: src/main/solved/BOJ/p1000/BOJ1091.java
import java.util.*;
import java.io.*;

class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] p = new int[N];
        int[] next = new int[N];
        int[] s = new int[N];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i=0; i<N; i++) {
            p[i] = Integer.parseInt(st.nextToken());
            next[i] = p[i];
        }

        st = new StringTokenizer(br.readLine());
        for (int i=0; i<N; i++) {
            s[i] = Integer.parseInt(st.nextToken());
        }

        int ans = 0;
        while(true) {
            if (validationCard(next)) {
                break;
            }
            shuffle(next, s);
            ans += 1;

            if (isSameOriginal(p, next)) {
                ans = -1;
                break;
            }
        }

        System.out.println(ans);
    }

    private static void shuffle(int[] next, int[] s) {
        int[] temp = next.clone();
        for (int i=0; i<next.length; i++) {
            temp[s[i]] = next[i];
        }

        for(int i=0; i<next.length; i++) {
            next[i] = temp[i];
        }
    }

    private static boolean validationCard(int[] next) {
        for (int i=0; i<next.length; i++) {
            if (next[i] != i%3) return false;
        }
        return true;
    }

    private static boolean isSameOriginal(int[] p, int[] next) {
        for (int i=0; i<next.length; i++) {
            if (p[i] != next[i]) {
                return false;
            }
        }
        return true;
    }
}

File: src/main/solved/BOJ/p1000/BOJ1091.kt
fun main() {
    val n = readLine()!!.toInt()
    val p = readLine()!!.split(" ").map { it.toInt() }.toList()
    val s = readLine()!!.split(" ").map { it.toInt() }.toList()
    val next = p.toMutableList()

    var ans = 0
    while(true) {
        if (isValidationOrder(next)) {
            break;
        }

        ans += 1
        shuffle(next, s)

        if (isOriginalSame(p, next)) {
            ans = -1;
            break;
        }
    }
    println(ans)
}

fun shuffle(next: MutableList<Int>, s: List<Int>) {
    val temp = arrayOfNulls<Int>(s.size)

    for (i in next.indices) {
        temp[s[i]] = next[i]
    }

    for (i in next.indices) {
        next[i] = temp[i] ?: -1
    }
}

fun isValidationOrder(next: MutableList<Int>): Boolean {
    for (i in 0 until next.size) {
        if (next[i] != i%3) return false
    }
    return true;
}

fun isOriginalSame(p: List<Int>, next: List<Int>): Boolean {
    for (i in p.indices) {
        if (p[i] != next[i]) {
            return false;
        }
    }
    return true;
}